tcms/testplans/tests/test_views.py

```diff
@@ -54,3 +56,167 @@ def test_view_tags_without_permissions(self):
+        remove_perm_from_user(self.user, self.add_testplan_permission)
+        response = self.client.get(self.location, follow=True)
+        self.assertRedirects(response, reverse('tcms-login') + '?next=' + self.location)
+
+    def test_post_with_no_perm_redirects_to_login(self):
+        remove_perm_from_user(self.user, self.add_testplan_permission)
+        response = self.client.post(self.location, self.request, follow=True)
+        self.assertRedirects(response, reverse('tcms-login') + '?next=' + self.location)
+
+
+class EditPlanViewTest(BasePlanTest):
+    test_plan_name = 'plan name'
+
+    def setUp(self):
+        super().setUp()
+        self.testplan_edit_url = reverse('plan-edit', args=[self.plan_id])
+
+        user_should_have_perm(self.user, perm='testplans.change_testplan')
+        self.user.is_superuser = False
+        self.user.save()
+
+        self.testplan_edit_data = {
+            'author': self.user.pk,
+            'product': self.product.pk,
+            'product_version': self.product_version.pk,
+            'type': self.plan_type.pk,
+            'name': self.test_plan_name,
+
+            'email_settings-0-auto_to_plan_author': 'on',
+            'email_settings-0-auto_to_case_owner': 'on',
+            'email_settings-0-auto_to_case_default_tester': 'on',
+            'email_settings-0-notify_on_case_update': 'on',
+            'email_settings-0-notify_on_plan_update': 'on',
+
+            'email_settings-0-plan': '',
+            'email_settings-0-id': self.test_plan.emailing.pk,
+            'email_settings-TOTAL_FORMS': '1',
+            'email_settings-INITIAL_FORMS': '1',
+            'email_settings-MIN_NUM_FORMS': '0',
+            'email_settings-MAX_NUM_FORMS': '1',
+            'is_active': True,
+        }
+
+    def test_show_testplan_edit_page(self):
+        response = self.client.get(self.testplan_edit_url)
+        self.assertContains(response, _('Edit TestPlan'))
+        assert_notify_buttons(self, response)
+
+    def test_edit_testplan_text_field(self):
+        edit_data = self.testplan_edit_data.copy()
+        new_text = 'Edited: {0}'.format(self.test_plan.text)
+        edit_data['text'] = new_text
+
+        response = self.client.post(self.testplan_edit_url, data=edit_data, follow=True)
+
+        self.assertContains(response, new_text)
+        self.test_plan.refresh_from_db()
+        self.assertEqual(new_text, self.test_plan.text)
+
+    def test_try_editting_testplan_with_invalid_form(self):
+        test_plan = TestPlanFactory(
+            author=self.user,
+            product=self.product,
+            product_version=self.product_version)
+
+        testplan_edit_url = reverse('plan-edit', args=[test_plan.pk])
+        edit_data = self.testplan_edit_data.copy()
+        edit_data['email_settings-0-auto_to_case_owner'] = 'invalid_value'
+
+        response = self.client.post(testplan_edit_url, data=edit_data, follow=True)
+
+        assert_notify_buttons(self, response)
```

Comment:
Now this doesn't properly assert the state after an error on the page.  B/c these buttons default to True/On you should change some of them to False/off in the post data and after the assert about the error message also assert that the buttons which were switched off are still off and the rest are still on. 

With this you are trying to verify this line of code:
```context['notify_formset'] = kwargs.get('notify_formset') or PlanNotifyFormSet()```

If we've got a bound form (even if it has errors) the page will show the error message and will preserve the button choices which have already been selected previously. If you remove the left part between `=` and `or` button choices will not be remembered between form submissions when there are errors.

Reply:
Are you sure about this, because form will not be submitted when there is an error, and i would expect these values to not change?

Reply:
Forms are always submitted to the backend view, even if they contain errors in their values. That's the whole point of forms, isn't it.

Perhaps you are thinking about 1 scenario and I am thinking about another one where both involve invalid forms.

Reply:
**Case 1 - When there is an invalid value on the main form:**
`notify_formset` is not preserved, except `is_active`. Its value is preserved, probably because it is also a member of the main_form.

I am not sure if this is an intended behaviour but when there is an invalid main form,  `kwargs.get('notify_formset') ` returns None and it reconstructs the `notify_formset` from `PlanNotifyFormSet(instance=self.object)` with default on/checked values.

**Case 2 - When there is an invalid value on notify form:**
`notify_formset` comes with the invalid value until here but on `rendered_content` after `render_to_response` it returns back to the default value, where all the notify values are on(checked).

`return self.render_to_response(self.get_context_data(notify_formset=notify_formset))`

A similar behavour also exist on NewTestPlan class.

![vokoscreen-2020-05-26_14-00-18](https://user-images.githubusercontent.com/8937774/82893848-57f88800-9f5a-11ea-924e-0b47f7124350.gif)

IMO, asserting to default notify formset values is sufficient.

Reply:
> A similar behavour also exist on NewTestPlan class.

It is the same underlying form and logic, that's why.

Reply:
For Case 1 - there is a bug in the code.

You need the following patch:
```
@@ -85,7 +87,11 @@ class Edit(UpdateView):
             return super().form_valid(form)
 
         # taken from FormMixin.form_invalid()
-        return self.render_to_response(self.get_context_data(notify_formset=notify_formset))
+        return self.render_to_response(self.get_context_data(form=form, notify_formset=notify_formset))
+
+    def form_invalid(self, form):
+        notify_formset = PlanNotifyFormSet(self.request.POST, instance=self.object)
+        return self.render_to_response(self.get_context_data(form=form, notify_formset=notify_formset))
```

When the main form is invalid we don't execute the `form_valid()` method at all so nobody tells `get_context` that there is a second form and that's why it reconstructs it without keeping the existing selections.

The same goes for `form_valid` - if the notify_form happens to be invalid we show the page again, this time keeping the notify_form but not the main one. You already have a scenario for this so before applying the patch if you try to assert one of the fields from the main form if should fail (I expect the value to be what comes from the DB, not what the user entered). You are doing something like that for the is_active field but strangely enough it didn't fail.

Reply:
> But tests have passed, both before and after applying the patch.

Yes because the current tests do not assert on the states of the on/off buttons when there is an error in the main form. Moreover to make sure that the patch really works at least one of these buttons should be OFF so that there is an expected difference from the default behavior.

Reply:
I am not sure if i am doing everyhting right, but even i apply the patch, the returned content is the default response with default 'on' values. By the way i have already defined the notify_formset values as 'off' by default on SetUpTestData - edit_data.

![image](https://user-images.githubusercontent.com/8937774/83327572-88576380-a285-11ea-83cd-f97c5573f63e.png)

I suspect from render_to_response method. Seems like everything returns to its default 'on' value when its rendered.

Reply:
I don't see the patch applied in your pull request.

Reply:
Do you want me to include the patch in this PR? I thought that is going to be considered on a separate issue.

Reply:
You can't write a correct test if the underlying code is not working correctly. Include the patch so we can move forward with the test.